### PR TITLE
[SCB-1017]add os cpu net info in the metrics with linux os

### PR DESCRIPTION
**Pull request by heyile**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/MetricsBootstrap.java [hunk 1]
```diff
@@ -41,13 +47,37 @@
...
+
   public void start(CompositeRegistry globalRegistry, EventBus eventBus) {
     this.globalRegistry = globalRegistry;
     this.eventBus = eventBus;
+    cpuNode = CpuNode.getINSTANCE();
     this.executorService = Executors.newScheduledThreadPool(1,
         new ThreadFactoryBuilder()
             .setNameFormat("spectator-poller-%d")
             .build());
+    if (!System.getProperty("os.name").toLowerCase().startsWith("win")) {
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/MetricsBootstrap.java]: try org.apache.commons.lang3.SystemUtils#IS_OS_LINUX

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/util/SystemOsUtil.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.servicecomb.foundation.metrics.publish.spectator.CpuNode;
+import org.apache.servicecomb.foundation.metrics.publish.spectator.NetNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SystemOsUtil {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SystemOsUtil.class);
+
+  private static final String linuxCpuCmd = "cat /proc/stat";
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/util/SystemOsUtil.java]: just read file directly

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/spectator/CpuNode.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.servicecomb.foundation.metrics.publish.spectator;
+
+import java.text.NumberFormat;
+
+public class CpuNode {
+
+  private static CpuNode INSTANCE = new CpuNode();
```
Comment (wujimin) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/spectator/CpuNode.java]: why need a global instance?

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/MetricsBootstrap.java (hunk 1 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/MetricsBootstrap.java]: ok

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/spectator/CpuNode.java (hunk 3 above)
Comment (heyile) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/spectator/CpuNode.java]: as I only need a single CpuNode instance, I was going to use the Singleton Pattern of java

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java [hunk 4]
```diff
@@ -148,7 +148,7 @@ public static void run() {
         for (String metricLine : metricLines) {
           if (!metricLine.startsWith("#")) {
             String[] metricKeyAndValue = metricLine.split(" ");
-            if (!metricKeyAndValue[0].startsWith("jvm")) {
+            if (!metricKeyAndValue[0].startsWith("jvm") && !metricKeyAndValue[0].startsWith("servicecomb_os")) {
```
Comment (wujimin) [inline demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java]: rename "servicecomb_os" to "os"

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java [hunk 5]
```diff
@@ -47,6 +53,8 @@
 
   public static final String ENABLED = "servicecomb.metrics.publisher.defaultLog.enabled";
 
+  public static final String METRICS_OS_ENABLED = "servicecomb.metrics.publisher.defaultLog.printOs.enabled";
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: why need this flag?

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java [hunk 6]
```diff
@@ -102,14 +111,67 @@ protected void printLog(List<Meter> meters) {
...
+        .get()) {
+      return;
+    }
+    MeasurementNode OsNode = tree.findChild(OsStatisticsMeter.OS_STATISTICS_NAME);
+    if (OsNode != null && !OsNode.getMeasurements().isEmpty()) {
+      MeasurementNode netNode = OsNode.findChild("net");
+      MeasurementNode cpuNode = OsNode.findChild("cpu");
+      if (cpuNode != null && !cpuNode.getMeasurements().isEmpty()) {
+        double rate = cpuNode.findChild("allCpu", OsCpuMeter.tagCpuRate.value()).summary();
+        String ratePercentStr = "calc ...";
```
Comment (wujimin) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: why "calc ..."?
why NumberFormat ?

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java (hunk 5 above)
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: ok

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java (hunk 6 above)
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: just give a tip... NumberFormat have removed

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java [hunk 7]
```diff
@@ -102,14 +111,67 @@ protected void printLog(List<Meter> meters) {
...
+            receiveRateStr = getRealTimeNetStr(receiveRate);
+          }
+          sb.append(String.format("    %-29s %-42s %s\n",
+              iNode.getName(), sendRateStr, receiveRateStr));
+        }
+      }
+    }
+  }
+
+  private String getRealTimeNetStr(Double netBytes) {
```
[the remark this answers is not in the document]
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: ok

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java (hunk 6 above)
Comment (heyile) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/DefaultLogPublisher.java]: I have removed the tip

File: foundations/foundation-common/src/test/java/org/apache/servicecomb/foundation/common/net/TestNetUtils.java [hunk 8]
```diff
@@ -108,6 +108,48 @@ public void testCanTcpListenYes() throws IOException {
...
+    Assert.assertEquals("1023.999 PB",
+        NetUtils.humanReadableBytes(1024L * 1024 * 1024 * 1024 * 1024 * 1024 - 1024L * 1024 * 1024 * 1024));
+
+    Assert.assertEquals("1.000 EB", NetUtils.humanReadableBytes(1024L * 1024 * 1024 * 1024 * 1024 * 1024));
+    Assert.assertEquals("1.001 EB",
+        NetUtils.humanReadableBytes(1024L * 1024 * 1024 * 1024 * 1024 * 1024 + 1024L * 1024 * 1024 * 1024 * 1024));
+    Assert.assertEquals("8.000 EB", NetUtils.humanReadableBytes(Long.MAX_VALUE));
+  }
+
+  public static void main(String[] args) {
```
Comment (wujimin) [inline foundations/foundation-common/src/test/java/org/apache/servicecomb/foundation/common/net/TestNetUtils.java]: remove main function

**Pull request merged**